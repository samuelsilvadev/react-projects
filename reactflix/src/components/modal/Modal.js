import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';

import { Button, Container, CloseStyled } from './Modal.style';

const Modal = ({ children, className, description, onClose, title,  }) => {
	const containerRef = useRef(null);

	const _handleKeyPress = (event) => {
		if (event.key === 'Escape') {
			onClose();
		}
	};

	const _setFocusOnModal = () => {
		if (containerRef && containerRef.current) {
			containerRef.current.focus();
		}
	};
	
	useEffect(() => {
		const focusedElBeforeOpen = document.activeElement;
		document.addEventListener('keydown', _handleKeyPress);
		_setFocusOnModal();

		return () => {
			focusedElBeforeOpen.focus();
			document.removeEventListener('keydown', _handleKeyPress);
		};
	}, []);

	return (
		<Container
			ref={ containerRef }
			role="dialog"
			aria-labelledby={ title }
			aria-describedby={ description }
			className={ className }
			data-enzyme-id="modal">
			<Button
				type="button"
				onClick={ onClose }
				data-enzyme-id="modal-close-button">
				<CloseStyled
					aria-label="Close Modal"
					data-enzyme-id="modal-close-icon" />
			</Button>
			{ children }
		</Container>
	);
};

Modal.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
	description: PropTypes.string,
	onClose: PropTypes.func.isRequired,
	title: PropTypes.string.isRequired,
};

Modal.defaultProps = {
	className: '',
	description: '',
};

export default Modal;
