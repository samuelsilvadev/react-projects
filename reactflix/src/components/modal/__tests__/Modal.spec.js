import React from 'react';
import { shallow } from 'enzyme';

import Modal from '../Modal';

describe('<Modal />', () => {
	it('should render correctly with button, close icon and correct children', () => {
		const requiredProps = {
			title: 'modal',
			onClose: jest.fn(),
		};
		const tree = shallow(
			<Modal { ...requiredProps }>
				<div data-enzyme-id="modal-children" />
			</Modal>
		);
		const $button = tree.find('[data-enzyme-id="modal-close-button"]');
		const $icon = tree.find('[data-enzyme-id="modal-close-icon"]');
		const $children = tree.find('[data-enzyme-id="modal-children"]');
		
		expect(tree.prop('aria-labelledby')).toBe(requiredProps.title);
		expect($icon.prop('aria-label')).toBe('Close Modal');

		expect($button).toBeDefined();
		expect($icon).toBeDefined();
		expect($children).toBeDefined();
		expect(tree).toMatchSnapshot();
	});

	it('should call `onClose` when click in close button', () => {
		const requiredProps = {
			title: 'modal',
			onClose: jest.fn(),
		};
		const tree = shallow(
			<Modal { ...requiredProps }>
				<div data-enzyme-id="modal-children" />
			</Modal>
		);
		const $button = tree.find('[data-enzyme-id="modal-close-button"]');
		
		$button.simulate('click');

		expect(requiredProps.onClose).toHaveBeenCalledTimes(1);
	});
});
