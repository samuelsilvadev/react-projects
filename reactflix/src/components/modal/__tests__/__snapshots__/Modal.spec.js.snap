// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render correctly with button, close icon and correct children 1`] = `ShallowWrapper {}`;
